Safe pruning of the segmented raft log.

maxsumrall opened the pull request:
- When attemping to prune, the pruner will not remove files which
  have open readers on the file. This behavior is unchanged but
  clarified by a test.
- The RaftLogShipper will attempt to send the requested entry.
  If that entry is pruned before the raft log shipper can send it,
  it will send the oldest entry it has, which will likely not be
  accepted by the follower and cause the follower to request again.
  The second request will be seen after pruning has happened,
  and the follower will be notified that a snapshot is needed.
- Rename the setting "CoreEdgeClusterSetting.raft_log_pruning"
  to "CoreEdgeClusterSetting.raft_log_pruning_strategy" with default value "keep_all".
- Removed RaftLogCompactedException. There is no need to differentiate
  between pruned indexes and indexes that do not yet exist.
- Properly parse prune strategy for PhysicalRaftLog.
- SegmentRaftLog EntryStore is refactored to EntryCursor.
- Cluster now implements AutoClosable.

Commit:
Safe pruning of the segmented raft log.

- When attemping to prune, the pruner will not remove files which
have open readers on the file. This behavior is unchanged but
clarified by a test.

- The RaftLogShipper will attempt to send the requested entry.
If that entry is pruned before the raft log shipper can send it,
it will send the oldest entry it has, which will likely not be
accepted by the follower and cause the follower to request again.
 The second request will be seen after pruning has happened,
and the follower will be notified that a snapshot is needed.

- Rename the setting "CoreEdgeClusterSetting.raft_log_pruning"
to "CoreEdgeClusterSetting.raft_log_pruning_strategy" with default value "keep_all".

- Removed RaftLogCompactedException. There is no need to differentiate
 between pruned indexes and indexes that do not yet exist.

- Properly parse prune strategy for PhysicalRaftLog.

- SegmentRaftLog EntryStore is refactored to EntryCursor.

- Cluster now implements AutoClosable.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/pruning/ThresholdConfigParser.java:
```diff
@@ -50,14 +50,16 @@ public static ThresholdConfigValue parse( String configValue )
+            case "true":
+                return ThresholdConfigValue.NO_PRUNING;
+            case "keep_none":
+            case "false":
```
is there any good reason to keep true/false?

digitalstain replied:
Only for consistency with the tx log configuration. Maybe a weak argument, and to be honest i'd like to see true/false be replaced with keep_all/keep_none in kernel as well. I think it is easier to understand.

davidegrohmann replied:
Oh.. now I get it, it is a shared configuration. It makes sense to be retro compatible then.

davidegrohmann commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/shipping/RaftLogShipper.java:
```diff
@@ -59,7 +57,7 @@
  * Each instance handles a single follower and acts on events and associated state updates originating
  * within the main raft state machine.
- *
+ * <p>
```
if we can avoid the `<p>` it would be great

davidegrohmann commented on enterprise/core-edge/src/test/java/org/neo4j/coreedge/scenarios/CorePruningIT.java:
```diff
@@ -70,6 +70,30 @@ public void actuallyDeletesTheFiles() throws Exception
+        {
+            coreGraphDatabase = cluster.coreTx( ( db, tx ) -> {
+                createData( db, 1 );
+                //tx.success();
```
??

davidegrohmann commented on enterprise/core-edge/src/test/java/org/neo4j/coreedge/scenarios/CoreToCoreCopySnapshotIT.java:
```diff
@@ -153,6 +154,71 @@ public void shouldBeAbleToDownloadAfterPruning() throws Exception
+
+            thread.start();
+
+            Thread.sleep( pruneFrequencyMs * 100 );
```
that's a long sleep...

maxsumrall replied:
Yes, to make sure that we get into the situation where a follower rejoins and needs pruned entries. We should refactor it to check the log files and see when that is rather than just waiting 10 seconds, but I would like to do that in a later PR.

davidegrohmann replied:
I would rather do it now, so we can avoid breaking builds in CI due to timing issue.

Commit:
Improved test for catching up rejoined follower.

The improved test explicitly checks the logs that have accumulated
and also the logs which have been pruned and coordinates the
shutdown and startup of the follower accordingly.

digitalstain commented on enterprise/core-edge/src/test/java/org/neo4j/coreedge/scenarios/CoreToCoreCopySnapshotIT.java:
```diff
@@ -164,59 +168,45 @@ public void shouldBeAbleToDownloadToRejoinedInstanceAfterPruning() throws Except
         int pruneFrequencyMs = 100;
         coreParams.put( raft_log_pruning_strategy.name(), "keep_none" );
         coreParams.put( raft_log_pruning_frequency.name(), pruneFrequencyMs + "ms" );
+        int numberOfTransactons = 100;
```
transactons. tehehehe... :)

digitalstain commented on enterprise/core-edge/src/test/java/org/neo4j/coreedge/scenarios/CoreToCoreCopySnapshotIT.java:
```diff
@@ -233,4 +223,53 @@ static void createData( GraphDatabaseService db, int size )
+
+    private String[] getLogFileNames( CoreGraphDatabase clusterMember )
+    {
+        File clusterDir = new File( clusterMember.getStoreDir(), CLUSTER_STATE_DIRECTORY_NAME );
```
It'd be more portable (and readable) if we used FileNames here.

digitalstain commented on enterprise/core-edge/src/test/java/org/neo4j/coreedge/scenarios/CoreToCoreCopySnapshotIT.java:
```diff
@@ -233,4 +223,53 @@ static void createData( GraphDatabaseService db, int size )
+    private Integer getOldestLogIdOn( CoreGraphDatabase clusterMember ) throws TimeoutException
+    {
+        String[] logsBefore = getLogFileNames( clusterMember );
+        return Integer.parseInt( logsBefore[0].replace( "raft.log.", "" ) );
```
this filename to version-as-int conversion would belong better in FileNames.

digitalstain commented:
@maxsumrall rebase needed
